### PR TITLE
Add MySQL session type

### PR DESCRIPTION
**Pull request by cgranleese-r7**
This PR adds a new session type specific to the MySQL protocol. This session _should_ behave as you might expect any other session in framework to behave with respect to interacting with the session i.e. `sessions -1` to drop into that sessions context and from there you could run MySQL specific commands like a generic query.

Example:
image

## Verification
Target: `docker run -it --rm -e MYSQL_ROOT_PASSWORD='password' -p 3306:3306 mariadb:latest`

- [ ] Start `msfconsole`
- [ ] `use scanner/mysql/mysql_login`
- [ ] Ensure this module now opens a session
- [ ] Interact with the session and ensure it works as other Metasploit Framework sessions do
- [ ] Run command dispatcher commands (e.g. `help`) while within the session and ensure all commands listed in the output work as described
- [ ] **Verify** that queries can be ran with both the `query` command and from within the raw shell after using the `shell` command
- [ ] **Verify** that modules can be ran from with sessions

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -60,6 +60,7 @@ PATH
       rb-readline
       recog
       redcarpet
+      reline
```
Comment (cgranleese-r7) [inline Gemfile.lock:63]: Remove as was landed as part off another PR.

File: lib/msf/base/sessions/mysql.rb [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    end
+  end
+
+  def type
+    self.class.type
+  end
+
+  #
+  # @return [String] The type of the session
+  #
```
Comment (cgranleese-r7) [inline lib/msf/base/sessions/mysql.rb:49]:
```suggestion
```

File: lib/msf/base/sessions/mysql.rb [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  def process_autoruns(datastore)
+    ['InitialAutoRunScript', 'AutoRunScript'].each do |key|
+      next if datastore[key].nil? || datastore[key].empty?
+
+      args = Shellwords.shellwords(datastore[key])
+      print_status("Session ID #{session.sid} (#{session.tunnel_to_s}) processing #{key} '#{datastore[key]}'")
+      session.execute_script(args.shift, *args)
+    end
+  end
+
```
Comment (cgranleese-r7) [inline lib/msf/base/sessions/mysql.rb:42]: Comment
```
  #
  # @return [String] The type of the session
  #
  ```

File: lib/msf/base/sessions/mysql.rb [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    @info = "MySQL #{datastore['USERNAME']} @ #{client.socket.peerinfo}"
+  end
+
+  def process_autoruns(datastore)
+    ['InitialAutoRunScript', 'AutoRunScript'].each do |key|
+      next if datastore[key].nil? || datastore[key].empty?
+
+      args = Shellwords.shellwords(datastore[key])
+      print_status("Session ID #{session.sid} (#{session.tunnel_to_s}) processing #{key} '#{datastore[key]}'")
+      session.execute_script(args.shift, *args)
```
Comment (cgranleese-r7) [inline lib/msf/base/sessions/mysql.rb:39]:
```suggestion
      self.execute_script(args.shift, *args)
```

Potentially needs changed.

File: lib/rex/post/mysql/ui/console/command_dispatcher/client.rb [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            until finished
+              begin
+                raw_query = ::Reline.readmultiline('SQL >> ', use_history = true) do |multiline_input|
+                  finished = stop_words.include?(multiline_input.split.last)
+                  !multiline_input.split.last.end_with?('\\')
+                end
+              rescue ::Interrupt
+                finished = true
+              ensure
+                ::Reline.prompt_proc = prompt_proc_before
```
Comment (cgranleese-r7) [inline lib/rex/post/mysql/ui/console/command_dispatcher/client.rb:89]: Check out Postgres changes

File: lib/rex/post/mysql/ui/console/command_dispatcher/client.rb [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+            query = args.join(' ').to_s
+            print_status("Sending statement: '#{query}'...")
+            result = client.query(query) || []
+
+            table = format_result(result)
+            print_line(table.to_s)
+          end
+
+          alias cmd_sql cmd_query
```
Comment (cgranleese-r7) [inline lib/rex/post/mysql/ui/console/command_dispatcher/client.rb:147]: Potentially these can be removed

File: lib/rex/post/mysql/ui/console/command_dispatcher/modules.rb [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+          # Modules
+          #
+          def name
+            'Modules'
+          end
+
+          def cmd_run_help
+            print_line 'Usage: Modules'
+            print_line
+            print_line 'Run a module.'
```
Comment (cgranleese-r7) [inline lib/rex/post/mysql/ui/console/command_dispatcher/modules.rb:42]: Worth checking each help and adding usage examples when appropriate.

File: metasploit-framework.gemspec [hunk 8]
```diff
@@ -244,4 +244,7 @@ Gem::Specification.new do |spec|
   # Do not use this to process untrusted PNG files! This is only to be used
   # to generate PNG files, not to parse untrusted PNG files.
   spec.add_runtime_dependency 'chunky_png'
+
+  # Needed for multiline REPL support for interactive SQL sessions
```
Comment (cgranleese-r7) [inline metasploit-framework.gemspec:248]: Remove

File: modules/auxiliary/scanner/mysql/mysql_login.rb [hunk 9]
```diff
@@ -5,13 +5,14 @@
 
 require 'metasploit/framework/credential_collection'
 require 'metasploit/framework/login_scanner/mysql'
+require 'rex/post/mysql'
```
Comment (dwelch-r7) [inline modules/auxiliary/scanner/mysql/mysql_login.rb:8]: is this needed?

File: modules/auxiliary/scanner/mysql/mysql_login.rb [hunk 10]
```diff
@@ -149,6 +175,23 @@ def int_version(str)
     end
   end
 
+  # @param [Metasploit::Framework::LoginScanner::Result] result
+  # @param [::Mysql] client
+  # @return [Msf::Sessions::MYSQL]
+  def session_setup(result, client)
+    return unless (result && client)
 
+    rstream = client
```
Comment (dwelch-r7) [inline modules/auxiliary/scanner/mysql/mysql_login.rb:184]: is this correct? I don't think rstream and client are supposed to be the same thing, pretty sure rstream is expected to be a Rex::Socket and then client would be your MySQL::Client?

**Review by dwelch-r7 — COMMENTED** [review]
Having another MySQL module made to use the session you've added in this PR would be helpful for testing

File: lib/msf/core/post/common.rb [hunk 11]
```diff
@@ -47,6 +49,8 @@ def rport
       session.session_port
     when 'postgresql'
       session.port
+    when 'mysql'
+      session.port
```
[the remark this answers is not in the document]
Comment (cgranleese-r7) [reply lib/msf/core/post/common.rb:53]: Updated this to combine the cases where values match. Do you think we should look into opening a ticket to update meterpreter and shell sessions adopt the same API? Just assuming we wouldn't want to do it as part of this PR.
```suggestion
    when 'postgresql', 'mysql'
      session.port
```

File: modules/auxiliary/scanner/mysql/mysql_login.rb (hunk 9 above)
Comment (cgranleese-r7) [reply modules/auxiliary/scanner/mysql/mysql_login.rb:8]: No seems it wasn't. I think I had added this when trying to get things working and forgot to remove it. Good catch 👍

File: lib/msf/core/post/common.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (dwelch-r7) [reply lib/msf/core/post/common.rb:53]: definitely not as part of this PR no, I'll look into making an issue for it, this change seems good though

[the remark this answers is not in the document]
Comment (cgranleese-r7) [reply lib/msf/core/post/common.rb:53]: Cool, thanks!

File: lib/msf/base/sessions/mysql.rb [hunk 12]
```diff
@@ -0,0 +1,136 @@
+# -*- coding: binary -*-
+
+require 'rex/post/mysql'
+
+class Msf::Sessions::MYSQL
```
[the remark this answers is not in the document]
Comment (cgranleese-r7) [reply lib/msf/base/sessions/mysql.rb:5]: Synced up on a call and got this updated. Cheers for the help @dwelch-r7

File: lib/metasploit/framework/login_scanner/mysql.rb [hunk 13]
```diff
@@ -64,8 +68,15 @@ def attempt_login(credential)
...
+          if mysql_conn
             result_options[:status] = Metasploit::Model::Login::Status::SUCCESSFUL
+
+            # This module no long owns the socket, return it as proof so the calling context can perform additional operations
+            # Additionally assign values to nil to avoid closing the socket etc automatically
+            if use_client_as_proof
+              result_options[:proof] = mysql_conn
+              mysql_conn = nil
+            end
           end
```
Comment (adfoster-r7) [inline lib/metasploit/framework/login_scanner/mysql.rb:80]: This is leaking file descriptors; similar to the mssql PR https://github.com/rapid7/metasploit-framework/pull/18696#issuecomment-1918220512

lsof:

```
ruby    19006 user   14u  IPv4 0x3f021c11117b81af      0t0                 TCP localhost:51805->localhost:pyrrho (ESTABLISHED)
ruby    19006 user   15u  IPv4 0x3f021c11123a104f      0t0                 TCP 192.168.123.1:51859->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   16u  IPv4 0x3f021c111239fa0f      0t0                 TCP 192.168.123.1:51860->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   17u  IPv4 0x3f021c11117a81af      0t0                 TCP 192.168.123.1:51871->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   18u  IPv4 0x3f021c11117a6b6f      0t0                 TCP 192.168.123.1:51872->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   19u  IPv4 0x3f021c11117a552f      0t0                 TCP 192.168.123.1:51874->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   20u  IPv4 0x3f021c11117a4a0f      0t0                 TCP 192.168.123.1:51875->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   21u  IPv4 0x3f021c1111ad1e2f      0t0                 TCP 192.168.123.1:51853->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   22u  IPv4 0x3f021c1111ad130f      0t0                 TCP 192.168.123.1:51854->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   23u  IPv4 0x3f021c1111acf1af      0t0                 TCP 192.168.123.1:51856->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   24u  IPv4 0x3f021c1111acfccf      0t0                 TCP 192.168.123.1:51857->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   25u  IPv4 0x3f021c11123a47ef      0t0                 TCP 192.168.123.1:51862->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   26u  IPv4 0x3f021c111116d7ef      0t0                 TCP 192.168.123.1:51863->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   27u  IPv4 0x3f021c111116ab6f      0t0                 TCP 192.168.123.1:51865->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   28u  IPv4 0x3f021c111116cccf      0t0                 TCP 192.168.123.1:51866->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   29u  IPv4 0x3f021c11117a8ccf      0t0                 TCP 192.168.123.1:51868->192.168.123.1:mysql (ESTABLISHED)
ruby    19006 user   30u  IPv4 0x3f021c11117a768f      0t0                 TCP 192.168.123.1:51869->192.168.123.1:mysql (ESTABLISHED)
```

Can we use the same approach as the postgres client to ensure any open sockets are closed:

https://github.com/rapid7/metasploit-framework/blob/15d22d2807a0ef9fe10599728cc92166195e9744/lib/metasploit/framework/login_scanner/postgres.rb#L76-L89

File: lib/msf/base/sessions/mysql.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (adfoster-r7) [reply lib/msf/base/sessions/mysql.rb:5]: > Synced up on a call and got this updated.

Looks like it wasn't updated? 👀 

```
class Msf::Sessions::MYSQL
```

File: modules/auxiliary/scanner/mysql/mysql_login.rb (hunk 9 above)
Comment (adfoster-r7) [reply modules/auxiliary/scanner/mysql/mysql_login.rb:8]: Looks like this is back again? Is the latest code pushed up here @cgranleese-r7 ?

Comment (cgranleese-r7) [reply modules/auxiliary/scanner/mysql/mysql_login.rb]: The changes hadn't been pushed up just yet. Sorry about the confusion.

File: lib/msf/base/sessions/mysql.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (cgranleese-r7) [reply lib/msf/base/sessions/mysql.rb]: The changes hadn't been pushed up just yet. Sorry about the confusion.

File: modules/auxiliary/scanner/mysql/mysql_login.rb (hunk 10 above)
Comment (cgranleese-r7) [reply modules/auxiliary/scanner/mysql/mysql_login.rb]: Yes, I have updated that to be the rex socket. In regards to `nil` working it was being handled as part of `start_session` where `my_session.rstream` was being passed.

File: lib/metasploit/framework/login_scanner/mysql.rb (hunk 13 above)
Comment (cgranleese-r7) [reply lib/metasploit/framework/login_scanner/mysql.rb:80]: Got this updated and working. Good catch! 👍

**Comment by cgranleese-r7** [comment]
Had some test failing with the recent changes:

```
52 examples, 5 failures

Failed examples:

rspec ./spec/lib/metasploit/framework/login_scanner/mysql_spec.rb:59 # Metasploit::Framework::LoginScanner::MySQL#attempt_login when the attempt is unsuccessful due to connection timeout returns a result with a status of Metasploit::Model::Login::Status::UNABLE_TO_CONNECT
rspec ./spec/lib/metasploit/framework/login_scanner/mysql_spec.rb:47 # Metasploit::Framework::LoginScanner::MySQL#attempt_login when the attempt is unsuccessful due to connection refused returns a result with a status of Metasploit::Model::Login::Status::UNABLE_TO_CONNECT
rspec ./spec/lib/metasploit/framework/login_scanner/mysql_spec.rb:83 # Metasploit::Framework::LoginScanner::MySQL#attempt_login when the attempt is unsuccessful due to not being allowed to connect from this host returns a result with a status of Metasploit::Model::Login::Status::UNABLE_TO_CONNECT
rspec ./spec/lib/metasploit/framework/login_scanner/mysql_spec.rb:71 # Metasploit::Framework::LoginScanner::MySQL#attempt_login when the attempt is unsuccessful due to operation timeout returns a result with a status of Metasploit::Model::Login::Status::UNABLE_TO_CONNECT
rspec ./spec/lib/metasploit/framework/login_scanner/mysql_spec.rb:39 # Metasploit::Framework::LoginScanner::MySQL#attempt_login when the attempt is successful returns a result object with a status of Metasploit::Model::Login::Status::SUCCESSFUL
```

I believe four failures were due to the existing API being designed to not return a `Login::Status::INCORRECT`, at least that's going by the tests, so I have removed that.
```
          else
            result_options[:status] = Metasploit::Model::Login::Status::INCORRECT
```

Other failure was due to the new `mysql_conn.close` to fix the leaking file descriptors. I updated the test to what they hopefully need to be 🤞 

Would appreciate someone else double checking that though!

File: modules/auxiliary/scanner/mysql/mysql_login.rb [hunk 14]
```diff
@@ -53,7 +66,7 @@ def run_host(ip)
         )
 
         scanner = Metasploit::Framework::LoginScanner::MySQL.new(
-            host: ip,
+            host: rhost,
```
Comment (adfoster-r7) [inline modules/auxiliary/scanner/mysql/mysql_login.rb:69]: It looks like this should be `ip` still

**Comment by adfoster-r7** [comment]
## Release Notes

Updates the `auxiliary/scanner/mysql/mysql_login.rb` module to include a new `CreateSession` option that opens an interactive session. This functonality is currently behind a feature flag which can be enabled with `features set mysql_session_type true`
